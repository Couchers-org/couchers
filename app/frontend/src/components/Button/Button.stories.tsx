// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Meta, Story } from "@storybook/react/types-6-0";
import * as React from "react";

import Button from "./Button";
// @ts-ignore

export default {
  title: "Components/Button",
  component: Button,
  argTypes: {
    backgroundColor: { control: "color" },
  },
} as Meta;

const Template: Story<any> = (args) => (
  <>
    <div style={{ display: "flex" }}>
      <Button {...args}>Click</Button>
    </div>
  </>
);

function wait(milliSeconds: number) {
  return new Promise((resolve) => setTimeout(resolve, milliSeconds));
}

export const Simple = Template.bind({});
Simple.args = {};

export const Loading = Template.bind({});
Loading.args = { loading: true };

export const AsyncOnClick = Template.bind({});
AsyncOnClick.args = { onClick: () => wait(1e3) };
