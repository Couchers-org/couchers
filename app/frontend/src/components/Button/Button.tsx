import {
  makeStyles,
  Button as MuiButton,
  ButtonProps,
  CircularProgress,
  Box,
} from "@material-ui/core";
import React, { ElementType, useState } from "react";
import classNames from "classnames";

const useStyles = makeStyles({
  root: {},
});

//type generics required to allow component prop
//see https://github.com/mui-org/material-ui/issues/15827
type AppButtonProps<D extends ElementType = "button", P = {}> = ButtonProps<
  D,
  P
> & {
  loading?: boolean;
};

export default function Button<D extends ElementType = "button", P = {}>({
  children,
  disabled,
  className,
  loading,
  onClick,
  ...otherProps
}: AppButtonProps<D, P>) {
  const [waiting, setWaiting] = useState(false);
  const classes = useStyles();
  async function asyncOnClick(event: any) {
    try {
      setWaiting(true);
      await onClick(event);
    } finally {
      setWaiting(false);
    }
  }
  return (
    <MuiButton
      {...otherProps}
      onClick={onClick && asyncOnClick}
      disabled={disabled ? true : loading || waiting}
      className={classNames(classes.root, className)}
    >
      {children}
      {(loading || waiting) && (
        <Box marginLeft="8px">
          <CircularProgress />
        </Box>
      )}
    </MuiButton>
  );
}
