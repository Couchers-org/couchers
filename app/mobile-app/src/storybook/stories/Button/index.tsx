import React from "react";
import PropTypes from "prop-types";
import { TouchableHighlight } from "react-native";

export default function Button({ onPress, children }: {onPress: any, children: any}) {
  return <TouchableHighlight onPress={onPress}>{children}</TouchableHighlight>;
}

Button.defaultProps = {
  children: null,
  onPress: () => {},
};

Button.propTypes = {
  children: PropTypes.node,
  onPress: PropTypes.func,
};
