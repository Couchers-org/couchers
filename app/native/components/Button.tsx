import { Pressable, StyleProp, StyleSheet, ViewStyle } from "react-native";
import { ThemedText } from "./ThemedText";

interface ButtonProps {
  title: string;
  filled?: boolean;
  onPress?: () => void;
  disabled?: boolean;
  style?: StyleProp<ViewStyle>;
}

export default function Button({
  title,
  filled = false,
  onPress,
  disabled = false,
  style,
}: ButtonProps) {
  return (
    <Pressable
      style={[
        styles.button,
        filled ? styles.filledButton : styles.transparentButton,
        disabled ? styles.disabledButton : null,
        style,
      ]}
      onPress={onPress}
      disabled={disabled}
    >
      <ThemedText
        type="button"
        style={[
          styles.buttonText,
          filled ? styles.filledButtonText : styles.transparentButtonText,
        ]}
      >
        {title}
      </ThemedText>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  button: {
    padding: 10,
    borderRadius: 5,
    alignItems: "center",
    marginBottom: 10,
  },
  filledButton: {
    backgroundColor: "#00A398",
  },
  disabledButton: {
    backgroundColor: "#E6EBEA",
  },

  transparentButton: {
    backgroundColor: "transparent",
    borderWidth: 1,
    borderColor: "#00A398",
  },
  buttonText: {
    fontWeight: "bold",
  },
  filledButtonText: {
    color: "white",
  },
  transparentButtonText: {
    color: "#00A398",
  },
});
