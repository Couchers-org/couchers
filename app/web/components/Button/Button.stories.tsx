import { Meta, Story } from "@storybook/react";
import * as React from "react";

import Button, { AppButtonProps } from "./Button";

export default {
  component: Button,
  title: "Components/Simple/Button",
} as Meta;

const Template: Story<AppButtonProps> = (args) => (
  <>
    <Button {...args}>Click</Button>
    <Button {...args}>Ok</Button>
    <Button {...args}>Longer label button</Button>
    <Button {...args} variant="text">
      Text button
    </Button>
    <Button {...args} variant="outlined">
      Outline button
    </Button>
  </>
);

export const Default: Story<AppButtonProps> = (args) => <Button {...args} />;
Default.args = {
  children: "Click me",
  variant: "contained",
  color: "primary",
};

export const Outlined: Story<AppButtonProps> = (args) => <Button {...args} />;
Outlined.args = {
  ...Default.args,
  variant: "outlined",
};

export const Text: Story<AppButtonProps> = (args) => <Button {...args} />;
Text.args = {
  ...Default.args,
  variant: "text",
};

function wait(milliSeconds: number) {
  return new Promise((resolve) => setTimeout(resolve, milliSeconds));
}

export const Simple = Template.bind({});
Simple.args = {};

export const Loading = Template.bind({});
Loading.args = { loading: true };

export const AsyncOnClick = Template.bind({});
AsyncOnClick.args = { onClick: () => wait(1e3) };
