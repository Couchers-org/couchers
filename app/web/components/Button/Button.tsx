import { Button as MuiButton, ButtonProps, useTheme } from "@material-ui/core";
import * as Sentry from "@sentry/react";
import classNames from "classnames";
import React, { ElementType } from "react";
import { useIsMounted, useSafeState } from "utils/hooks";
import makeStyles from "utils/makeStyles";

import CircularProgress from "../CircularProgress";

const useStyles = makeStyles((theme) => ({
  contained: {
    borderRadius: theme.shape.borderRadius,
    boxShadow: "0px 0px 5px rgba(0, 0, 0, 0.25)",
  },
  loading: {
    height: theme.typography.button.fontSize,
  },
  spinner: {
    position: "absolute",
    bottom: 0,
    left: 0,
    right: 0,
    top: 0,
    margin: "auto",
  },
  root: {
    minHeight: `calc(calc(${theme.typography.button.lineHeight} * ${
      theme.typography.button.fontSize
    }) + ${theme.typography.pxToRem(12)})`, //from padding
  },
}));

//type generics required to allow component prop
//see https://github.com/mui-org/material-ui/issues/15827
export type AppButtonProps<
  D extends ElementType = "button",
  P = {}
> = ButtonProps<D, P> & {
  loading?: boolean;
};

export default function Button<D extends ElementType = "button", P = {}>({
  children,
  disabled,
  className,
  loading,
  onClick,
  variant = "contained",
  color = "primary",
  ...otherProps
}: AppButtonProps<D, P>) {
  const isMounted = useIsMounted();
  const [waiting, setWaiting] = useSafeState(isMounted, false);
  const classes = useStyles();
  const theme = useTheme();
  async function asyncOnClick(event: any) {
    try {
      setWaiting(true);
      await onClick(event);
    } catch (e) {
      Sentry.captureException(e);
    } finally {
      setWaiting(false);
    }
  }
  if (variant !== "contained" && color !== "primary") {
    throw new Error("Only contained buttons should have color.");
  }
  return (
    <MuiButton
      {...otherProps}
      onClick={onClick && asyncOnClick}
      disabled={disabled ? true : loading || waiting}
      className={classNames(classes.root, className, {
        [classes.contained]: variant === "contained",
      })}
      variant={variant}
      color={variant === "contained" ? color : undefined}
    >
      {(loading || waiting) && (
        <CircularProgress
          size={theme.typography.button.fontSize}
          className={classes.spinner}
        />
      )}
      {children}
    </MuiButton>
  );
}
