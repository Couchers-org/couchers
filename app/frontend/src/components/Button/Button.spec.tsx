import React from "react";
import renderer from "react-test-renderer";
import Button from "./Button";

describe("Button", () => {
  it("should match snapshot", () => {
    const tree = renderer.create(<Button />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
