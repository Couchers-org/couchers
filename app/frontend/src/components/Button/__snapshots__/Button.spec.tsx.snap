// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-root-1"
  disabled={false}
  onBlur={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  <span
    className="MuiButton-label"
  />
</button>
`;
